ironfish/src/wallet/scanner/walletScanner.ts

```diff
@@ -0,0 +1,247 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import type { Wallet } from '../wallet'
+import { BufferMap } from 'buffer-map'
+import { Config } from '../../fileStores'
+import { Logger } from '../../logger'
+import { Mutex } from '../../mutex'
+import { RpcClient } from '../../rpc'
+import { AsyncUtils, BufferUtils, HashUtils } from '../../utils'
+import { DecryptedNote } from '../../workerPool/tasks/decryptNotes'
+import { HeadValue } from '../walletdb/headValue'
+import {
+  RemoteChainProcessor,
+  WalletBlockHeader,
+  WalletBlockTransaction,
+} from './remoteChainProcessor'
+import { ScanState } from './scanState'
+
+export class WalletScanner {
+  readonly logger: Logger
+  readonly nodeClient: RpcClient | null
+  readonly wallet: Wallet
+
+  maxQueueSize: number
+  state: ScanState | null = null
+  lock: Mutex = new Mutex()
+
+  constructor(options: {
+    logger: Logger
+    nodeClient: RpcClient | null
+    wallet: Wallet
+    maxQueueSize: number
+    config: Config
+  }) {
+    this.logger = options.logger
+    this.nodeClient = options.nodeClient
+    this.wallet = options.wallet
+    this.maxQueueSize = options.maxQueueSize
+  }
+
+  get running(): boolean {
+    return !!this.state
+  }
+
+  async abort(): Promise<void> {
+    if (this.state) {
+      await this.state.abort()
+    }
+  }
+
+  async wait(): Promise<void> {
+    if (this.state) {
+      await this.state.wait()
+    }
+  }
+
+  async scan({
+    start,
+    end,
+    force,
+  }: {
+    start?: HeadValue | null
+    end?: HeadValue | null
+    force?: boolean
+  } = {}): Promise<ScanState | null> {
+    if (this.wallet.listAccounts().length === 0) {
+      return null
+    }
+
+    if (this.running && !force) {
+      this.logger.debug('Skipping Scan, already scanning.')
+      return null
+    }
+
+    if (this.running && force) {
+      this.logger.debug('Aborting scan in progress and starting new scan.')
+      await this.abort()
+    }
+
+    await this.state?.wait()
+    const unlock = await this.lock.lock()
```

Comment:
This lock plus the void async code below feels a bit confusing -- it might be nice to either have a test that demonstrates why this is here, or I wonder if we could rework it somehow so `this.scan` acts as a lock.

Kind of thinking of something like this, but it might be pushing too much logic into ScanState
```
await this.state?.wait()
this.state = ScanState.init(async () => {
  // set up scanning
})
```

Reply:
I'll do this immediately in the next PR. We'll most likely move all this logic into ScanState and rename it to WalletScan.